File: cs/cslox/Parser/Expr.cs
using cslox.Scanning;

namespace cslox.Parser;

interface IExprVisitor<T>
{
	T VisitAssignExpr(Assign expr);
	T VisitBinaryExpr(Binary expr);
	T VisitCallExpr(Call expr);
	T VisitGetExpr(Get expr);
	T VisitGroupingExpr(Grouping expr);
	T VisitLiteralExpr(Literal expr);
	T VisitLogicalExpr(Logical expr);
	T VisitSetExpr(Set expr);
	T VisitUnaryExpr(Unary expr);
	T VisitVariableExpr(Variable expr);
}

abstract class Expr
{
	public abstract T Accept<T>(IExprVisitor<T> visitor);
}

class Assign : Expr
{
	public Assign(Token name, Expr value)
	{
		Name = name;
		Value = value;
	}

	public Token Name { get; }
	public Expr Value { get; }

	public override T Accept<T>(IExprVisitor<T> visitor)
	{
		return visitor.VisitAssignExpr(this);
	}
}

class Binary : Expr
{
	public Binary(Expr left, Token @operator, Expr right)
	{
		Left = left;
		Operator = @operator;
		Right = right;
	}

	public Expr Left { get; }
	public Token Operator { get; }
	public Expr Right { get; }

	public override T Accept<T>(IExprVisitor<T> visitor)
	{
		return visitor.VisitBinaryExpr(this);
	}
}

class Call : Expr
{
	public Call(Expr callee, Token paren, List<Expr> arguments)
	{
		Callee = callee;
		Paren = paren;
		Arguments = arguments;
	}

	public Expr Callee { get; }
	public Token Paren { get; }
	public List<Expr> Arguments { get; }

	public override T Accept<T>(IExprVisitor<T> visitor)
	{
		return visitor.VisitCallExpr(this);
	}
}

class Get : Expr
{
	public Get(Expr @object, Token name)
	{
		Object = @object;
		Name = name;
	}

	public Expr Object { get; }
	public Token Name { get; }

	public override T Accept<T>(IExprVisitor<T> visitor)
	{
		return visitor.VisitGetExpr(this);
	}
}

class Grouping : Expr
{
	public Grouping(Expr expression)
	{
		Expression = expression;
	}

	public Expr Expression { get; }

	public override T Accept<T>(IExprVisitor<T> visitor)
	{
		return visitor.VisitGroupingExpr(this);
	}
}

class Literal : Expr
{
	public Literal(object? value)
	{
		Value = value;
	}

	public object? Value { get; }

	public override T Accept<T>(IExprVisitor<T> visitor)
	{
		return visitor.VisitLiteralExpr(this);
	}
}

class Logical : Expr
{
	public Logical(Expr left, Token @operator, Expr right)
	{
		Left = left;
		Operator = @operator;
		Right = right;
	}

	public Expr Left { get; }
	public Token Operator { get; }
	public Expr Right { get; }

	public override T Accept<T>(IExprVisitor<T> visitor)
	{
		return visitor.VisitLogicalExpr(this);
	}
}

class Set : Expr
{
	public Set(Expr @object, Token name, Expr value)
	{
		Object = @object;
		Name = name;
		Value = value;
	}

	public Expr Object { get; }
	public Token Name { get; }
	public Expr Value { get; }

	public override T Accept<T>(IExprVisitor<T> visitor)
	{
		return visitor.VisitSetExpr(this);
	}
}

class Unary : Expr
{
	public Unary(Token @operator, Expr right)
	{
		Operator = @operator;
		Right = right;
	}

	public Token Operator { get; }
	public Expr Right { get; }

	public override T Accept<T>(IExprVisitor<T> visitor)
	{
		return visitor.VisitUnaryExpr(this);
	}
}

class Variable : Expr
{
	public Variable(Token name)
	{
		Name = name;
	}

	public Token Name { get; }

	public override T Accept<T>(IExprVisitor<T> visitor)
	{
		return visitor.VisitVariableExpr(this);
	}
}

File: tree-walk-interpreter/go/internal/parser/expr.go
package parser

import (
	"github.com/kashifsoofi/go-lox/internal/scanner"
)

type ExprVisitor interface {
	VisitAssignExpr(expr *Assign) interface{}
	VisitBinaryExpr(expr *Binary) interface{}
	VisitCallExpr(expr *Call) interface{}
	VisitGetExpr(expr *Get) interface{}
	VisitGroupingExpr(expr *Grouping) interface{}
	VisitLiteralExpr(expr *Literal) interface{}
	VisitLogicalExpr(expr *Logical) interface{}
	VisitSetExpr(expr *Set) interface{}
	VisitSuperExpr(expr *Super) interface{}
	VisitThisExpr(expr *This) interface{}
	VisitUnaryExpr(expr *Unary) interface{}
	VisitVariableExpr(expr *Variable) interface{}
}

type Expr interface {
	Accept(v ExprVisitor) interface{}
}

type Assign struct {
	Name  *scanner.Token
	Value Expr
}

func NewAssign(name *scanner.Token, value Expr) *Assign {
	return &Assign{
		Name:  name,
		Value: value,
	}
}

func (expr *Assign) Accept(v ExprVisitor) interface{} {
	return v.VisitAssignExpr(expr)
}

type Binary struct {
	Left     Expr
	Operator *scanner.Token
	Right    Expr
}

func NewBinary(left Expr, operator *scanner.Token, right Expr) *Binary {
	return &Binary{
		Left:     left,
		Operator: operator,
		Right:    right,
	}
}

func (expr *Binary) Accept(v ExprVisitor) interface{} {
	return v.VisitBinaryExpr(expr)
}

type Call struct {
	Callee    Expr
	Paren     *scanner.Token
	Arguments []Expr
}

func NewCall(callee Expr, paren *scanner.Token, arguments []Expr) *Call {
	return &Call{
		Callee:    callee,
		Paren:     paren,
		Arguments: arguments,
	}
}

func (expr *Call) Accept(v ExprVisitor) interface{} {
	return v.VisitCallExpr(expr)
}

type Get struct {
	Object Expr
	Name   *scanner.Token
}

func NewGet(object Expr, name *scanner.Token) *Get {
	return &Get{
		Object: object,
		Name:   name,
	}
}

func (expr *Get) Accept(v ExprVisitor) interface{} {
	return v.VisitGetExpr(expr)
}

type Grouping struct {
	Expression Expr
}

func NewGrouping(expression Expr) *Grouping {
	return &Grouping{
		Expression: expression,
	}
}

func (expr *Grouping) Accept(v ExprVisitor) interface{} {
	return v.VisitGroupingExpr(expr)
}

type Literal struct {
	Value interface{}
}

func NewLiteral(value interface{}) *Literal {
	return &Literal{
		Value: value,
	}
}

func (expr *Literal) Accept(v ExprVisitor) interface{} {
	return v.VisitLiteralExpr(expr)
}

type Logical struct {
	Left     Expr
	Operator *scanner.Token
	Right    Expr
}

func NewLogical(left Expr, operator *scanner.Token, right Expr) *Logical {
	return &Logical{
		Left:     left,
		Operator: operator,
		Right:    right,
	}
}

func (expr *Logical) Accept(v ExprVisitor) interface{} {
	return v.VisitLogicalExpr(expr)
}

type Set struct {
	Object Expr
	Name   *scanner.Token
	Value  Expr
}

func NewSet(object Expr, name *scanner.Token, value Expr) *Set {
	return &Set{
		Object: object,
		Name:   name,
		Value:  value,
	}
}

func (expr *Set) Accept(v ExprVisitor) interface{} {
	return v.VisitSetExpr(expr)
}

type Super struct {
	Keyword *scanner.Token
	Method  *scanner.Token
}

func NewSuper(keyword *scanner.Token, method *scanner.Token) *Super {
	return &Super{
		Keyword: keyword,
		Method:  method,
	}
}

func (expr *Super) Accept(v ExprVisitor) interface{} {
	return v.VisitSuperExpr(expr)
}

type This struct {
	Keyword *scanner.Token
}

func NewThis(keyword *scanner.Token) *This {
	return &This{
		Keyword: keyword,
	}
}

func (expr *This) Accept(v ExprVisitor) interface{} {
	return v.VisitThisExpr(expr)
}

type Unary struct {
	Operator *scanner.Token
	Right    Expr
}

func NewUnary(operator *scanner.Token, right Expr) *Unary {
	return &Unary{
		Operator: operator,
		Right:    right,
	}
}

func (expr *Unary) Accept(v ExprVisitor) interface{} {
	return v.VisitUnaryExpr(expr)
}

type Variable struct {
	Name *scanner.Token
}

func NewVariable(name *scanner.Token) *Variable {
	return &Variable{
		Name: name,
	}
}

func (expr *Variable) Accept(v ExprVisitor) interface{} {
	return v.VisitVariableExpr(expr)
}
